covid19_sfbayarea/data/contra_costa.py

```diff
@@ -0,0 +1,415 @@
+             for x, y in new_cases_data}
+    for x, y in total_cases_data:
+        day = api.parse_date(x['qNum'])
+        record = cases.get(day)
+        if not record:
+            record = {'date': day.isoformat()}
+            cases[day] = record
+
+        record['cumul_cases'] = y['qNum']
+
+    result = [cases[day] for day in sorted(cases.keys())]
+
+    # Sanity-check daily cases vs. totals
+    total = 0
+    for index, record in enumerate(result):
+        total += record['cases']
+        if record['cumul_cases'] != total:
+            raise FormatError(f'Sum of daily cases != cumul_cases at record {index} (date: {record["day"]})')
+
+    return result
+
+
+def get_timeseries_deaths(api: QlikClient) -> List[dict]:
+    data = get_chart_data(api, CHART_IDS['deaths'])
+
+    total = 0
+    results = []
+    # LTCF = Long Term Care Facility (e.g. nursing homes)
+    # LTCF deaths include both residents and staff.
+    for x, non_ltcf, ltcf in data:
+        day = api.parse_date(x['qNum'])
+        deaths = non_ltcf['qNum'] + ltcf['qNum']
+        total += deaths
+        results.append({
+            'date': day.isoformat(),
+            'deaths': deaths,
+            'cumul_deaths': total,
+            # TODO: consider adding these for counties that include them?
+            # 'ltcf_deaths': ltcf['qNum'],
+            # 'non_ltcf_deaths': non_ltcf['qNum'],
+        })
+
+    return results
+
+
+def get_timeseries_tests(api: QlikClient) -> List[dict]:
+    """
+    Calculate a timeseries of tests.
+
+    The dashboard has daily tests and daily total tests as separate charts, but
+    they don't quite agree. We use the daily total tests chart and calculate
+    daily tests by taking the difference between each day.
+
+    More detail: daily total tests starts earlier (but clearly not on the first
+    day there were tests, given the values). Daily tests starts later *and* has
+    ``0`` as the value for the first several days, despite the daily totals
+    having > 0 differences on those days. Once daily tests starts having values
+    it lines up with the differences between daily total tests entries, so it
+    seems like daily total tests is a more complete and accurate chart.
+    """
+    tests: Dict[date, Dict] = defaultdict(lambda: {
+        'tests': -1,
+        'positive': -1,
+        'negative': -1,
+        'pending': -1,
+        'cumul_tests': -1,
+        'cumul_pos': -1,
+        'cumul_neg': -1,
+        'cumul_pend': -1,
+        'positivity': -1
+    })
+
+    total_tests_data = get_chart_data(api, CHART_IDS['total_tests'])
+    for x, y in total_tests_data:
+        day = api.parse_date(x['qNum'])
+        tests[day]['cumul_tests'] = y['qNum']
+
+    # There's no chart with total positive tests, so we use the *rate* to
+    # calculate it. Note also that this is a rolling average, not the actual
+    # rate on the day, so this is not the most accurate. :(
```

Comment:
Well noted. How should we communicate these types of limitations on the front end?

Reply:
I tried to note this in the `meta_from_baypd` field, and I think for now that’s the best technical method we have. But it definitely seems like we might be able to structure this stuff better for display (maybe a `notes` field in each section of the data or something?) or could just hard-code similar information in the UI. Not sure. :\